### DISCUSSION TITLE
[Question] How to modifiy the gyro ouput data with fault injection?(I want to know about sensor data source)

### DISCUSSION THREAD

**Discussion by doltae09** [bitcraze/discussions#690] [Q&A] [answered]
Hello! I want to make sensor fault output in crazyflie-firmware.

The gyro data is put to 3EA c-file(1-"estimator_complementary.c" / 2-"controller_pid.c" / 3-"attitude_pid_controller.c") in complementary filter environment. (Crazyflie 2.1 onboard + z-ranger --> condition attached deck on crazyflie 2.1)

In this condition, I want to add the bias fault in "gyro.x".
Where is the gyro data source to send the data for 3EA c-file(1-"estimator_complementary.c" / 2-"controller_pid.c" / 3-"attitude_pid_controller.c")?
I want to add the gyro fault data that can apply to 3-EA c-file totally . That is a purpose.

**Comment by krichardsson** [comment]
Hi!

I'm not exactly sure what you mean and need some more information to provide a good answer.
As far as I understand, you want to inject an error into the gyro data. 
Is this something you want to do in the Crazyflie (firmware) or in simulation on a PC?
Also, I'm not sure what "3EA" means.
Maybe you can also tell us a bit about what you want to achieve, why you want to insert the error.

**Reply by doltae09** [reply]
Oh! I'm sorry that I didn't say detaily.
I am studying about IMU sensor fault condition.
I want to test on firmware level about fault injection of IMU.
Specially, I want to inject to gyro fault data in sensor data with working in controller & estimator.

Now I consider the experiment environments.
[1] Firmware Level : gyro fault injection coding
[2] PC cfclient : I can inject to param by coding [1]

So, firstly I make a example code for gyro fault injection right this (code location : "attitude_pid_controller.c").
![image](https://user-images.githubusercontent.com/79627957/235045102-08f1dfaa-eb30-4c73-a7df-01006637611e.png)

it is work well, but gyro data is used directly in controller & estimator in below deck platform
- Crazyflie 2.1 + z-ranger deck  --> estimator (complemetary filter)
In this case, How can inject fault data in gyro with estimator("estimator_complementary.c") and controller("attitude_pid_controller.c") together??

I think that this is root source of gyro sensor data for contoller("attitude_pid_controller.c") & estimator("estimator_complementary.c").
![image](https://user-images.githubusercontent.com/79627957/235046512-f1564640-00eb-4d89-a612-5bed5514b1f0.png)

Can you give me some guidelines??

**Reply by krichardsson** [reply, marked as answer]
Thanks for the clarification!

I think the easiest place to hook into is [in the stabilizer loop](https://github.com/bitcraze/crazyflie-firmware/blob/39cf2a2bbb80859d3cea23aacd419b21cd5d2a33/src/modules/src/stabilizer.c#L268)

`sensorsAcquire(&sensorData);` reads the sensor data, and after this you can modify the values in sensorData to your liking. This is the data that will be used by controllers and estimators later on. This way your error injection will work regardless of which estimator, controller or deck that is used.
